### ISSUE TITLE
Fails to cross-compile due to IFLA_* being mandatory

### ISSUE THREAD

**Issue by th0ma7** [arvidn/libtorrent#6792]
libtorrent version (or branch): RC_2_0 and 1.2.15

platform/architecture: i686 target (cross-compiled from x86_64 host) on Linux using Synology NAS DSM toolchain.

compiler and compiler version:  gcc 7.5 (and also 4.9 for older DSM)

Able to cross-compile on `armv7` (32-bit), `aarch64` (armv8, 64-bit) and `x86_64` for Synology Linux DSMv7 (as DSMv6 uses a too old gcc).  Although it fails i686 32-bit, qoriq (64-bit ppc), armv5, comcerto (armv7l)

Assocated PR: https://github.com/SynoCommunity/spksrc/pull/5180
See build output and logs here: https://github.com/SynoCommunity/spksrc/pull/5180/checks

Log of interest with failure on i686:
[build-evansport-7.0.log](https://github.com/arvidn/libtorrent/files/8405532/build-evansport-7.0.log)

This is a follow-up of #6788 now that build is working in almost all cases.

**Comment by th0ma7** [comment]
I'm also getting this same error on other archs, not only i686.  Would it be possible to add a flag to disable them as needed or check if they are supported instead of having to manually patch the code such as we currently do https://github.com/SynoCommunity/spksrc/pull/4153/files#diff-31d2f52b1915b81e0a1f69a31c02dfe9b70143cd0b1cc86dd2ba5b74930b19c4

**Comment by AllSeeingEyeTolledEweSew** [comment]
For x86 32-bit, that's generally tested by libtorrent's CI. Check some recent github actions runs for our setup.

I don't see a x86 32-bit build in the workflow actions you linked. In the build log, it looks like `address-model=64` is still passed, so maybe that's a source of trouble.

For all the other archs, we don't / can't cross-compile those in CI. So if you submit a patch to make it work and it doesn't break existing archs, arvid will likely just accept it.

**Comment by th0ma7** [comment]
> I don't see a x86 32-bit build in the workflow actions you linked. In the build log, it looks like `address-model=64` is still passed, so maybe that's a source of trouble.

Using Synology toolchain it's called `evansport` and is using `address-model=32`

As for a patch... I can try but really this ain't that much my cup of tea...

**Comment by th0ma7** [comment]
> I don't see a x86 32-bit build in the workflow actions you linked. In the build log, it looks like `address-model=64` is still passed, so maybe that's a source of trouble.

Revisiting this and you are right, all my 32-bit archs are being plagued with `address-model=64` while the boost `user-config.jam` does uses <address-model>"32".  How can I ensure libtorrent uses the user-config.jam OR can I pass in argument address-model=32 as needed?

**Comment by AllSeeingEyeTolledEweSew** [comment]
`setup.py` detects whether to pass `address-model=(32|64)` based on `sys.maxsize`. So if it's passing `address-model=64`, it means you're using a 64-bit python installation, which seems like an error if you're trying to build a 32-bit wheel.

You could also look at our `cibuildwheel` setup. I already tested that it works for `i686`, `ppc64le`, and `s390x`. I left those out of our publishing workflow just to save CI time, but you could fork the repo and enable them in `.github/workflows/cibuildwheel.yml`. I don't think it supports the 32-bit ARM archs though.

**Comment by th0ma7** [comment]
That might explain why...  Correct me if I'm wrong but here's my theory:
We pre-build a "native" python-3.10 that can run on the `x86_64` host which is then used to generate pure-python wheels.  As this is a cross-compiled wheel from a `x86_64` host to `armv5`/`7`/`8`/`ppc` targets, crossed-compiled binaries aren't functional on the host and are not part of the default `PATH`.  Therefore I fear that, in a cross-compiling environment, `sys.maxsize` will always points to the host which it turn is 64-bit being wrong for many archs.

If I get this right I would then absolutely need a method to either pass `--b2-args=address-model=32` or even better, having the ability to point to the boost build-time user configuration file `--b2-args=user-config=$(WORK_DIR)/boost_1_70_0/user-config.jam`.  On the other hand, the `BOOST_BUILD_PATH` environment variable already points to the root directory where the `user-config.jam` is located, so couldn't `setup.py` simply fetch it if available?

I haven't played with `cibuildwheel` setup, it's now next on my "to have a look" list.

**Comment by AllSeeingEyeTolledEweSew** [comment]
I didn't know about `crossenv`. At the time I rejiggered `setup.py` I thought python cross-compiling basically didn't exist. `setup.py`'s goal is just "customize b2 for the *target* python", and I'd understood that *target* python always meant *running* python, but I guess that's not the case for `crossenv`. It seems much more sane to have `setup.py` be aware of `crossenv`'s build target, rather than some other hacks. I'll take a look at this.

In the meantime `cibuildwheel` is certainly the fastest way to get  wheels for `i686` and `ppc64le`. Just fork, modify `cibuildwheel.yml`, and run the workflow. The wheels will be available as build artifacts (after a couple hours).

**Comment by AllSeeingEyeTolledEweSew** [comment]
So `crossenv` monkeypatches a python installation to look like a target installation. So the philosophy that "`setup.py` should build for the invoking python" is still valid.

[The standard library documentation says that `sys.maxsize` is the source of truth for whether the running python is 32- or 64-bit](https://docs.python.org/3/library/platform.html#platform.architecture), which is why we use this test today. But [`crossenv` doesn't patch this](https://github.com/benfogle/crossenv/blob/d757aa274962ab9e866e8bf81180ab24e9f4e723/crossenv/scripts/sys-patch.py.tmpl). So this is really a bug in `crossenv`.

More info on other possible sources of bitness:

1. Most sources of *parsed* bitness (unambiguously indicates 32- or 64-bit) reflect the *host*'s bitness in 32-on-64 installations. That is, if I install `python3:i386` on 64-bit ubuntu, these functions return 64-bit. I don't want to use these functions for `address-model`, because I want such an environment to build 32-bit libtorrent. These include `platform.machine()`, `platform.uname()`, `platform.platform()`.
2. There are many sources of *unparsed* bitness, where e.g. we'd have to parse `x86_64-linux-gnu` to mean 64-bit. These sources are mostly correctly patched by `crossenv` and could be used to discover its target build, but I don't want to use them, because it's far too error-prone. These sources include `sysconfig`, `distutils.sysconfig`, and the various system paths.
3. [`platform.architecture()`](https://docs.python.org/3/library/platform.html#platform.architecture) unambiguously returns `"32bit"`/`"64bit"`. It returns `"32bit"` in 32-on-64-bit installations. The function [uses `struct.calcsize()`](https://github.com/python/cpython/blob/c9d41bcd68dbe339396523e931904930a87819b9/Lib/platform.py#L664-L667) as a fallback. `crossenv` doesn't patch either of these functions.

**Comment by th0ma7** [comment]
> So `crossenv` monkeypatches a python installation to look like a target installation. So the philosophy that "`setup.py` should build for the invoking python" is still valid.

Which is exactly what I'm calling in my use-case (e.g. crossenv python310).

> [The standard library documentation says that `sys.maxsize` is the source of truth for whether the running python is 32- or 64-bit](https://docs.python.org/3/library/platform.html#platform.architecture), which is why we use this test today. But [`crossenv` doesn't patch this](https://github.com/benfogle/crossenv/blob/d757aa274962ab9e866e8bf81180ab24e9f4e723/crossenv/scripts/sys-patch.py.tmpl). So this is really a bug in `crossenv`.

I've opened-up an issue on the matter here:  https://github.com/benfogle/crossenv/issues/97

> More info on other possible sources of bitness:

Going back to my previous point, since probing python for bitness does not work currently, could you use the boost `user-config.jam` which sets-it?  Here's an example for ppc (qoriq) where `<address-model>"32"` is set:
```
$ cat user-config.jam 
using gcc : 4.9.3 : /home/spksrc/master/spksrc/toolchain/syno-qoriq-6.1/work/powerpc-e500v2-linux-gnuspe/bin/powerpc-e500v2-linux-gnuspe-g++ : <address-model>"32" <cflags>"-I/home/spksrc/master/spksrc/toolchain/syno-qoriq-6.1/work/powerpc-e500v2-linux-gnuspe/powerpc-e500v2-linux-gnuspe/sysroot/usr/include -mcpu=8548 -mhard-float -mfloat-gprs=double -I/home/spksrc/master/spksrc/spk/python310/work-qoriq-6.1/install//var/packages/python310/target/include " <cxxflags>"-I/home/spksrc/master/spksrc/toolchain/syno-qoriq-6.1/work/powerpc-e500v2-linux-gnuspe/powerpc-e500v2-linux-gnuspe/sysroot/usr/include -mcpu=8548 -mhard-float -mfloat-gprs=double -I/home/spksrc/master/spksrc/spk/python310/work-qoriq-6.1/install//var/packages/python310/target/include -Wno-deprecated-declarations -I/home/spksrc/master/spksrc/spk/python310/work-qoriq-6.1/install/var/packages/python310/target/include/python3.10 -fpermissive" <linkflags>"-L/home/spksrc/master/spksrc/toolchain/syno-qoriq-6.1/work/powerpc-e500v2-linux-gnuspe/powerpc-e500v2-linux-gnuspe/sysroot/lib -mcpu=8548 -mhard-float -mfloat-gprs=double -L/home/spksrc/master/spksrc/spk/python310/work-qoriq-6.1/install//var/packages/python310/target/lib -Wl,--rpath-link,/home/spksrc/master/spksrc/spk/python310/work-qoriq-6.1/install//var/packages/python310/target/lib -Wl,--rpath,/var/packages/python310/target/lib " <link>"shared" ; 
project : requirements <cxxflags> ;
```

**Comment by AllSeeingEyeTolledEweSew** [comment]
@arvidn I think the `IFLA_*` problems are a red herring for wrong `address-model=`, so this should still be open

@th0ma7 `setup.py` just invokes `b2` in `bindings/python`. That will pick up any `user-config.jam` within `b2`'s search paths, so there should be no need to pass `--user-config=...`.

**Comment by th0ma7** [comment]
> @arvidn I think the `IFLA_*` problems are a red herring for wrong `address-model=`, so this should still be open

@AllSeeingEyeTolledEweSew the `address-model` is a totally separated issue not related to `IFLA_*`.  The patchset still applies.

> @th0ma7 `setup.py` just invokes `b2` in `bindings/python`. That will pick up any `user-config.jam` within `b2`'s search paths, so there should be no need to pass `--user-config=...`.

I can confirm that it doesn't OR that I haven't found the proper method to pass the right argument for its search path.  And looking at the `setup.py` code it seems to only be searching for it through the host python call which, as discussed previously, is not covered by `crossenv`.

**Comment by AllSeeingEyeTolledEweSew** [comment]
> the `address-model` is a totally separated issue not related to `IFLA_*`. The patchset still applies.

I based that claim on the fact that I'd been able to run our `cibuildwheel` setup on `i686` and `ppc64le` without encountering `IFLA_*` constant issues.
 
> I can confirm that it doesn't OR that I haven't found the proper method to pass the right argument for its search path. And looking at the `setup.py` code it seems to only be searching for it through the host python call which, as discussed previously, is not covered by `crossenv`.

`setup.py` doesn't detect, touch or configure anything to do with `user-config.jam`.

`setup.py` creates its own `project-config.jam`. It does this to force b2 to build against the python installation which invoked `setup.py` (otherwise b2 will tend to use any random global python installation).

b2 has a lot of complex, implicit logic for how build properties are propagated, which gets layered on the complexity of a project's Jamfiles. Here, you want to override a property from the build request for *all* targets; I don't know how you do that, or if it's even possible.

I'd probably solve this problem by sticking a `b2` wrapper script in `$PATH` that mutates its arguments and dispatches to the real `b2`.

**Comment by th0ma7** [comment]
Which in turns goes back to the need of an ability to pass `--b2-args=address-model=32` when needed to circumvent `crossenv` inability to provide proper bitness, right?

**Comment by arvidn** [comment]
when configuring a cross compiler in `~/user-config.jam`, you can specify the `address-model`, and you'd probably want to specify `target-os` and maybe `architecture` as well.

Here's an example: https://blog.libtorrent.org/2017/02/building-libtorrent-for-c-h-i-p/

The configuration file that's auto generated by the python build script just configures the python library and include path.

**Comment by th0ma7** [comment]
Using a unique `~/user-config.jam` isn't possible.  Our framework builds in parallel multiple archs, each under their own sub-directoy, where we already generated a `user-config.jam` under their own boost build sub-directory (e.g. `BOOST_BUILD_PATH`)
```
.
├── work-88f6281-6.1
│   ├── boost_1_70_0
│   │   └── user-config.jam
│   ├── crossenv
│   ├── Python-3.10.4
│   └── ...
├── work-aarch64-6.1
├── work-aarch64-7.0
├── work-armv7-6.1
├── work-armv7-7.0
├── work-comcerto2k-7.0
├── work-evansport-6.1
├── work-evansport-7.0
├── work-hi3535-6.1
├── work-qoriq-6.1
├── work-x64-6.1
└── work-x64-7.0
    ├── boost_1_70_0
    │   └── user-config.jam
    ├── crossenv
    ├── Python-3.10.4
    └── ...
```
Here is an example of our generated file for armv5 which works perfectly fine for a boost build:
```
$ cat work-88f6281-6.1/boost_1_70_0/user-config.jam 
using gcc : 4.6.4 : /home/spksrc/master/spksrc/toolchain/syno-88f6281-6.1/work/arm-marvell-linux-gnueabi/bin/arm-marvell-linux-gnueabi-g++ : <address-model>"32" <cflags>"-I/home/spksrc/master/spksrc/toolchain/syno-88f6281-6.1/work/arm-marvell-linux-gnueabi/arm-marvell-linux-gnueabi/libc/usr/include -I/home/spksrc/master/spksrc/spk/python310/work-88f6281-6.1/install//var/packages/python310/target/include " <cxxflags>"-I/home/spksrc/master/spksrc/toolchain/syno-88f6281-6.1/work/arm-marvell-linux-gnueabi/arm-marvell-linux-gnueabi/libc/usr/include -I/home/spksrc/master/spksrc/spk/python310/work-88f6281-6.1/install//var/packages/python310/target/include -Wno-deprecated-declarations -I/home/spksrc/master/spksrc/spk/python310/work-88f6281-6.1/install/var/packages/python310/target/include/python3.10 -fpermissive" <linkflags>"-L/home/spksrc/master/spksrc/toolchain/syno-88f6281-6.1/work/arm-marvell-linux-gnueabi/arm-marvell-linux-gnueabi/libc/lib -L/home/spksrc/master/spksrc/spk/python310/work-88f6281-6.1/install//var/packages/python310/target/lib -Wl,--rpath-link,/home/spksrc/master/spksrc/spk/python310/work-88f6281-6.1/install//var/packages/python310/target/lib -Wl,--rpath,/var/packages/python310/target/lib " <link>"shared" ; 
project : requirements <cxxflags> ;
```
So conceptually we already have a `user-config.jam` ready and available for using where `<address-model>"32"` is set.  To me all is needed is either the ability to pass that config file OR bitness in parameter as `--b2-args=`.

**Comment by AllSeeingEyeTolledEweSew** [comment]
I think I missed what you were asking. You're saying `--b2-args=address-model=32` doesn't work for you currently? I believe it may need `--b2-args=address-model=32 --no-autoconf=address-model`.

**Comment by arvidn** [comment]
I believe it's also possible to pass:
```
--b2-args=toolset=gcc-464
```

Which would imply the correct address-model for your target, along with other things.

Looking at `setup.py`, it seems it always adds `address-model=` to the command line, based on the host python's `sys.maxsize`. I think this is probably the core of the issue.

**Comment by th0ma7** [comment]
> I believe it's also possible to pass:
> 
> ```
> --b2-args=toolset=gcc-464
> ```
> 
> Which would imply the correct address-model for your target, along with other things.

Odd but I'll give that a shot (wonder how this will turn out on gcc-7.5 use-cases).

> Looking at `setup.py`, it seems it always adds `address-model=` to the command line, based on the host python's `sys.maxsize`. I think this is probably the core of the issue.

@arvidn it's exactly that!

**Comment by AllSeeingEyeTolledEweSew** [comment]
> Looking at `setup.py`, it seems it always adds `address-model=` to the command line, based on the host python's `sys.maxsize`. I think this is probably the core of the issue.

@arvidn actually when you backported my `setup.py` changes to `RC_1_2` and merged them through, it looks like you added `--no-autoconf`, and I missed that on review. My intent was that `--b2-args=foo=bar` would imply `--no-autoconf=foo`. It's been in production for a bit now though so we should just keep it
